7.sc_timelapse_model/scripts/0.data_cleaning.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import pathlib
+import time
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import euclidean
+
+try:
+    cfg = get_ipython().config
+    in_notebook = True
+except NameError:
+    in_notebook = False
+if in_notebook:
+    from tqdm.notebook import tqdm
+else:
+    from tqdm import tqdm
+
+
+# In[2]:
+
+
+sc_profile_file_path = pathlib.Path(
+    "../../data/CP_scDINO_features/combined_CP_scDINO_norm_fs.parquet"
+).resolve(strict=True)
+endpoint_sc_profile_file_path = pathlib.Path(
+    "../../data/CP_feature_select/endpoints/features_selected_profile.parquet"
+).resolve(strict=True)
+sc_profile_df = pd.read_parquet(sc_profile_file_path)
+endpoint_sc_profile_df = pd.read_parquet(endpoint_sc_profile_file_path)
+endpoint_sc_profile_df["Metadata_Well_FOV"] = (
+    endpoint_sc_profile_df["Metadata_Well"].astype(str)
+    + "_"
+    + endpoint_sc_profile_df["Metadata_FOV"].astype(str)
+)
+print(endpoint_sc_profile_df.shape)
+sc_profile_df["Metadata_sc_unique_track_id"] = (
+    sc_profile_df["Metadata_Well"].astype(str)
+    + "_"
+    + sc_profile_df["Metadata_FOV"].astype(str)
+    + "_"
+    + sc_profile_df["Metadata_track_id"].astype(str)
+)
+sc_profile_df["Metadata_Well_FOV"] = (
+    sc_profile_df["Metadata_Well"].astype(str)
+    + "_"
+    + sc_profile_df["Metadata_FOV"].astype(str)
+)
+print(sc_profile_df.shape)
+sc_profile_df.head()
+
+
+# In[3]:
+
+
+# drop all nan values in the location columns
+endpoint_sc_profile_df = endpoint_sc_profile_df.dropna(
+    subset=["Metadata_Nuclei_Location_Center_X", "Metadata_Nuclei_Location_Center_Y"]
+)
+
+
+# In[4]:
+
+
+last_time_point_df = sc_profile_df.loc[
+    sc_profile_df["Metadata_Well_FOV"].isin(
+        endpoint_sc_profile_df["Metadata_Well_FOV"].unique()
+    )
+]
+
+
+# In[5]:
+
+
+print(len(last_time_point_df["Metadata_Well_FOV"].unique()))
+print(len(endpoint_sc_profile_df["Metadata_Well_FOV"].unique()))
+
+
+# In[6]:
+
+
+dict_of_sc_well_fovs = {}
+for well_fov in last_time_point_df["Metadata_Well_FOV"].unique():
+    dict_of_sc_well_fovs[well_fov] = last_time_point_df[
+        last_time_point_df["Metadata_Well_FOV"] == well_fov
+    ].copy()
+    # get only the last timepoint of the track
+
+    dict_of_sc_well_fovs[well_fov].reset_index(drop=True, inplace=True)
+dict_of_sc_well_fovs_endpoint = {}
+for well_fov in endpoint_sc_profile_df["Metadata_Well_FOV"].unique():
+    dict_of_sc_well_fovs_endpoint[well_fov] = endpoint_sc_profile_df[
+        endpoint_sc_profile_df["Metadata_Well_FOV"] == well_fov
+    ].copy()
+    dict_of_sc_well_fovs_endpoint[well_fov].reset_index(drop=True, inplace=True)
+
+
+# In[7]:
+
+
+start_time = time.time()
+
+
+# In[8]:
+
+
+for well_fov in tqdm(list(dict_of_sc_well_fovs.keys()), desc="Processing Well-FOVs"):
```

Comment:
Needs documentation - why is fuzzy matching necessary? Is this what the tracking is supposed to do? Unclear to me